src/core/or/circuitbuild.c

```diff
@@ -2580,7 +2584,24 @@ choose_good_middle_server(uint8_t purpose,
     return choice;
   }
```

Comment:
I think the sanest way for them to interact is by having the vanguards addon inspect MiddleNodes, and only choose HSLayer2Nodes from that set. Otherwise, the manpage should say that researchers must manually set both values to the same nodes if they want to test padding on vanguards-like circuits without using the addon.

Reply:
Documented in 74404f7f5aa2a73b96d2fa1e7bf89c23c10090ac